README.md
Paging
======

Break long pages into smaller ones automatically by words/characters limit, or
by means of a customizable pagination tag. For example:

    First page here.
    <!--pagebreak-->
    Second page here.
    <!--pagebreak-->
    More pages here.


Installation
------------

- Install this module using the official Backdrop CMS instructions at
  https://backdropcms.org/guide/modules

- Configure the general pagination settings under Administration >
  Configuration > Content authoring > Paging (admin/config/content/paging).

- Enable paging for the content types you want to use it on, and define specific
  settings at for each type, under Administration > Structure > Content types
  (admin/structure/types)


Documentation
-------------

Additional documentation is located in the Wiki:
https://github.com/backdrop-contrib/paging/wiki/Documentation

Issues
------

Bugs and Feature requests should be reported in the Issue Queue:
https://github.com/backdrop-contrib/paging/issues

Current Maintainers
-------------------

- Jen Lampton (https://github.com/jenlampton)
- Seeking additional maintainers

Credits
-------

- Original module written by Marco Scutari.
- Rewritten, considerably shortened & made more friendly by [Earl Miles](https://www.drupal.org/u/merlinofchaos).
- Yet again rewritten, extended, and maintained by [Gurpartap Singh](https://www.drupal.org/u/gurpartap-singh).
- Ported to Drupal 7 and then Backdrop CMS by [Jen Lampton](https://github.com/jenlampton)

License
-------

This project is GPL v2 software. See the LICENSE.txt file in this directory for
complete text.


CHANGELOG.txt

Paging 6.x-1.0, 2009-xx-xx
------------------------------
by Gurpartap Singh: UX: Displays a successful enabling message along with link to settings page, upon enabling the module.
#434652 by jcmarco: Remove @ from selectors.
#366862 by smacphail & Gurpartap Singh: node_teaser() isn't suitable when teaser break tag is present.
#383670 by Gurpartap Singh: Expose a "current page" variable for themes.
#438002 by Gurpartap Singh & RTH: Missing argument error on filter tips page.

Paging 6.x-1.0-beta3, 2009-01-27
------------------------------
#185032 by Gurpartap Singh: Paging block (shows a list of page names as links).
by Gurpartap Singh: Fixed empty names populating method
by Gurpartap Singh: Fixed page navigation block visibility.
by Gurpartap Singh: Disable page names UI when automatic paging is enabled.
by Gurpartap Singh: Remove paging filter settings in hook_uninstall().
#266373 by sun: Add configurable position for pager above and/or below content or custom pager output in theme.
by Gurpartap Singh: Added comments to the complete code for better understanding when revisiting.
by Gurpartap Singh: Added CHANGELOG.txt with complete change log till date.
by Gurpartap Singh: Added a new theme function to set select list for drop down pager theme function. This provides the ability to override it and set custom URLs(even clean urls).
by Gurpartap Singh: Added limitations topic to README.txt
by Gurpartap Singh: Markdown format for README.txt contents.
by Gurpartap Singh: Created a JavaScript theme function for drop down pager, to allow overriding it's behavior.
by Gurpartap Singh: Fixed array_fill method.
#45153 by SiliconMind: Do not paginate book module's print pages.
by Gurpartap Singh: Fixed error on 'add new comment' page displaying paginated node.
#346329 by wwalc, Gurpartap Singh: E_NOTICE compatibility.
#363396 by Gurpartap Singh: Paginate only for normal node rendering. Fixed several reported issues where other content fields went missing.
by Gurpartap Singh: Move setting of default value in drop down pager, from JavaScript to theme function in module. Reduces the code to be loaded client side.
by Gurpartap Singh: Remove first and last link from drop down pager. Include next and previous page's name along with next and previous links.
#329923 by Gurpartap Singh: AJAX load the pages. Uses nedjo's code proposed for AJAX pagers in core.
by Gurpartap Singh: Names can be specified anywhere in the body.
by Gurpartap Singh: Better method for wrapping pager and body for AJAX compatibility.
by Gurpartap Singh: Implement hook_content_extra_fields() to allow customized positioning of pager(s) using "Manage fields" under CCK.
#362966 by Gurpartap Singh: Fixed automatic paging in middle of tag.


Paging 6.x-1.0-beta2, 2008-11-08
------------------------------
#330057 by Gurpartap Singh: Page names interface bug + Drop down nav bug.
by Gurpartap Singh: Fixed parameters for pager and the drop down theme function
by john vandyk: Removed junk code.


Paging 6.x-1.0-beta1, 2008-10-17
------------------------------
#284076 by john vandyk: Code refinements
#91084 by Gurpartap Singh: Close to full page names support
by Gurpartap Singh: New drop down list pager navigation.
by Gurpartap Singh: Fixed the layout of drop down list
by Gurpartap Singh: Fixed description for drop down pager.
by Gurpartap Singh: New feature "Page names" along with a new administration UI.
by Gurpartap Singh: Paging for Drupal 6.


Paging 5.x-1.1, 2008-01-25
------------------------------
by Gurpartap Singh: Fixed critical bugs with automatic paging
#126365 by Gurpartap Singh: Automatic Paging
by Darren Oh: Used strpos instead of substr_count to increase speed.
#133961 by Darren Oh: Paging interfered with search.
by Darren Oh: Removed update index op from hook_nodeapi implementation (no longer needed).
by Darren Oh: Changed $node->pages_count to $node->page_count (better grammatically).
by Darren Oh: Added package tag to info file now that we're creating some associated modules.
#118294 by Darren Oh: Better text, text formatting and code indentation.


Paging 5.x-1.0, 2006-12-02
------------------------------
by Gurpartap Singh: Removed unnecessary css file: paging.css.
by Gurpartap Singh: Temporary fix for book navigation links.
by Gurpartap Singh: Settings form's fieldset label fixed.
by Gurpartap Singh: Added Drupal 5 compatibility.


Paging 4.7-1.x, 2006-04-02
------------------------------
#79130 by douggreen: New Paging Module that supports filters and CCK


Paging 4.7-1.x, 2006-04-02
------------------------------
by chx: Simplified and fixed.


Paging 4.6-1.x, 2006-04-02
------------------------------
by chx: Simplified and fixed.


Paging 4.6-1.x, 2006-03-14
------------------------------
#49814 by merlinofchaos: HTML tags eaten.
by merlinofchaos: Fix for comment pager collision.
by merlinofchaos: Fix for comment pager collision (part 2)


Paging 4.7-1.x, 2006-03-14
------------------------------
#51217 by merlinofchaos: comment pager problem.
#49814 by merlinofchaos: HTML tags being eaten.


Paging 4.7-1.x, 2006-02-19
------------------------------
#47324 by anders: Fix for filter format


Paging 4.6-1.x, 2006-01-22
------------------------------
by merlinofchaos: Backported from 4.7


Paging 4.7-1.x, 2006-01-19
------------------------------
#45153 by merlinofchaos: print.module problem


Paging 4.7-1.x, 2006-01-10
------------------------------
by merlinofchaos: Replaced <!-- page --> with <!--pagebreak--> for TinyMCE compatibility


Paging 4.7-1.x, 2006-01-04
------------------------------
by merlinofchaos: removed end ?> which is not necessary


Paging 4.7-1.x, 2005-12-14
------------------------------
by merlinofchaos: Removed extra CR at end of module.


Paging 4.7-1.x, 2005-12-13
------------------------------
by merlinofchaos: Rewrote paging module for 4.7. Now about 1/3 as long and uses Drupal's
pager.


Paging 4.6-1.x, 2005-12-01
------------------------------

admin/paging.admin.js
// $Id$

/**
 * Vertical tabs node form layout (aka drawers).
 * @author Bevan Rudge drupal.geek.nz
 * @author Joakim Stai
 * @license GPL 2
 */

$(document).ready(function(){
  // Turn all marked fieldsets into vertical tabs.
  $('.paging-fieldset').each(function(i){
    $('#paging-vertical-tabs > ul')
      .append('<li>' + $(this).find('.paging-enabled').parent().parent().html() + '<a href="#' + $(this).attr('id') + '">' + $(this).find('> legend').text() + '</a></li>')
      .find('> li').click(function() {
        $(this).find('> a').click();
      });
    $(this).removeClass('collapsible collapsed').appendTo('#paging-vertical-tabs');
    $(this).find('.paging-enabled').parent().parent().remove();
  });

  // Activate vertical tabs.
  $('#paging-vertical-tabs > ul').tabs({fx: { opacity: 'toggle', duration: 150 }, cookie: { expires: 1, path: '/' }});
  

  // Find the height of the heighest panel.
  var maxHeight = $.map($('#paging-vertical-tabs'),function(el){
    return $(el).height();
  }).sort(function(a, b) {
    return b - a;
  })[0];

  // Set the height of the container to a fixed size.
  $('#paging-vertical-tabs').height(maxHeight);

  $(".paging-method input[type=radio]").bind("change", paging_automatic_handler).each(paging_automatic_handler);
  $(".paging-pager input[type=radio]").bind("change", paging_pager_handler).each(paging_pager_handler);
});

function paging_automatic_handler(event) {
  var type = $(this).attr('name').substr(24);
  var checked = $('input[name=paging_automatic_method_' + type + ']:checked').val();
  $('.paging-chars-' + type + ', .paging-words-' + type).hide();
  $('input[name=paging_names_enabled_' + type + ']').removeAttr('disabled');

  if (checked == 1) {
    $('.paging-chars-' + type).fadeIn(500);
    $('input[name=paging_names_enabled_' + type + ']').removeAttr('checked');
    $('input[name=paging_names_enabled_' + type + ']').attr('disabled', 'disabled');
  }
  else if (checked == 2) {
    $('.paging-words-' + type).fadeIn(500);
    $('input[name=paging_names_enabled_' + type + ']').removeAttr('checked');
    $('input[name=paging_names_enabled_' + type + ']').attr('disabled', 'disabled');
  }
}

function paging_pager_handler(event) {
  var type = $(this).attr('name').substr(20);
  var checked = $('input[name=paging_pager_widget_' + type + ']:checked').val();
  $('.paging-pager-custom-' + type).parent().hide();

  if (checked == 'custom') {
    $('.paging-pager-custom-' + type).parent().fadeIn(500);
  }
}


paging.js
// $Id$

Number.prototype.ordinal = function () {
  return this + (
    (this % 10 == 1 && this % 100 != 11) ? '<sup>st</sup>' :
    (this % 10 == 2 && this % 100 != 12) ? '<sup>nd</sup>' :
    (this % 10 == 3 && this % 100 != 13) ? '<sup>rd</sup>' : '<sup>th</sup>'
  );
};

function paging_handle_names(pages) {
  var output = '';

  if (pages.length > 0) {
    for (var x = 0; x < pages.length; x++) {
      var title = $('#edit-title').val();
      var name = Drupal.t('!title - Page !number', {'!title': title, '!number': (x + 1)});
      pages[x] = (pages[x] || name);
      output += '<label for="edit-paging-title[' + x + ']">' + Drupal.t("Name of !number page: ", {'!number': (x + 1).ordinal()}) + '</label>'
        + "\t" + '<input type="text" class="form-text" value="' + pages[x] + '" size="20" name="edit-paging-title[' + x + ']" maxlength="255"/>' + "\n";
    }
    $('#paging-names-wrapper').show().each(function() {
      $(this).html('<fieldset class="form-item" id="paging-page-names" style="overflow: hidden;">'
      + '<legend class="">' + Drupal.t("Page names") + '</legend><div id="paging-names-inner" class="form-item"></div></fieldset>');
    });
    $('#paging-names-wrapper #paging-names-inner').html(output);
  }
  else {
    $('#paging-names-wrapper').hide();
  }
}

function paging_return_names() {
  var names = [];
  var i = 0;
  $('#paging-page-names').find('input[type=text]').each(function() {
    names[i] = $(this).val();
    i++;
  }).parents('fieldset').remove();

  return names;
}

function paging_insert_separator() {
  var separator = Drupal.settings.paging.separator;
  $('textarea#edit-body').each(function() {
    if (document.selection) {
      this.focus();
      document.selection.createRange().text = separator;
    }
    else if (this.selectionStart || this.selectionStart == '0') {
      var cursorPos = this.selectionEnd + separator.length;
      this.value = this.value.substring(0, this.selectionEnd) + separator + this.value.substring(this.selectionEnd);
      this.selectionStart = this.selectionEnd = cursorPos;
    }
    else {
      this.value = this.value + separator;
    }
    this.focus();
  }).trigger('click');

  return false;
}

Drupal.behaviors.paging = function(context) {
  // Handle the widget button to insert paging separator.
  $('textarea.teaser').each(function() {
    var include = $('#' + this.id.substring(0, this.id.length - 2) + 'include');
    var widget = Drupal.settings.paging.widget;

    if (widget == 1) {
    // Add paging separator string image.
      var path = Drupal.settings.basePath + Drupal.settings.paging.module_path;
      var string = Drupal.t('Insert page separator');
      var image = $('<div class="paging-button-wrapper"><img alt="" src="' + path + '/paging-separator.png" class="paging-separator-image" /></div>');
      $(include).parent().parent().before(image);
      $('img', image).attr('alt', string).attr('title', string).bind('click', paging_insert_separator);
    }
    else if (widget == 2) {
      // Add paging separator string button.
      var button = $('<div class="paging-button-wrapper"><input type="button" class="paging-separator-button" /></div>');
      $(include).parent().parent().before(button);
      $('input', button).val(Drupal.t('Insert page separator')).bind('click', paging_insert_separator);
    }
  });

  var page_names = typeof Drupal.settings.paging != 'undefined' ? Drupal.settings.paging.page_names : 0;
  if (page_names == 1) {
  // Page names interface is enabled.
    $('#edit-body-wrapper').after('<div id="paging-names-wrapper"></div>');

    $('textarea#edit-body').each(function() {
      var str = $(this).val();
      var match = str.match(/<!--pagenames:(.*?)-->/) || [];
      $(this).val(str.replace(/<!--pagenames:(.*?)-->/, ''));
      var pages = typeof match[1] != 'undefined' ? match[1].split('||') : [];
      paging_handle_names(pages);
    }).bind('click keyup blur', function() {
      var str = $(this).val();
      var separator = Drupal.settings.paging.separator;
      var count = 0;
      var current_names = paging_return_names();
      var pages = [(current_names[count] ? current_names[count] : '')];
      var pos = str.indexOf(separator);
      while (pos != -1) {
        count++;
        pages.push(current_names[count] ? current_names[count] : '');
        pos = str.indexOf(separator, pos + 1);
      }
      paging_handle_names(str.indexOf(separator) != -1 ? pages : 0);
    });

    $('#node-form').submit(function() {
      $('textarea#edit-body').each(function() {
        var names = paging_return_names();
        if (names.length > 0) {
          $(this).val('<!--pagenames:' + names.join('||') + '-->' + this.value);
        }
      });
    });
  }

  $('div.paging-pager-contents:not(.pager-processed)', context).each(function() {
    var $target = $(this);
    $target.addClass('pager-processed')
    // Process pager links.
    .find('ul.pager > li > a, table.paging-drop-down a')
    .each(function() {
      var ajaxPath = $(this).attr('href');
      $(this).click(function () {
        $link = $(this);
        $link.addClass('throbber');
        $.ajax({
          url: ajaxPath,
          type: 'GET',
          dataType: 'json',
          data: {'paging_json_request': $target.attr('id')},
          success: function(response) {
            $link.removeClass('throbber');
            if (response.content) {
              var $newContent = $(response.content);
              $target.replaceWith($newContent);
              Drupal.attachBehaviors($newContent.parent());
              var offset = $newContent.parent().offset().top;
              $('html,body').animate({scrollTop: offset}, 500);
            }
          },
          error: function() {
            $link.removeClass('throbber');
            alert(Drupal.t("An error occurred at @path.", {'@path': ajaxPath}));
          }
        });
        return false;
      });
    });
  });
};
